core/shared/src/main/scala/fs2/util/Concurrent.scala

```diff
@@ -0,0 +1,299 @@
+  }
+
+  /** Constructs a `Concurrent` instance from a supplied `Effect` instance and execution context. */
+  def mk[F[_]](F: Effect[F])(implicit ec: ExecutionContext): Concurrent[F] = {
+    final class MsgId
+    sealed abstract class Msg[A]
+    object Msg {
+      final case class Read[A](cb: Attempt[(A, Long)] => Unit, id: MsgId) extends Msg[A]
+      final case class Nevermind[A](id: MsgId, cb: Attempt[Boolean] => Unit) extends Msg[A]
+      final case class Set[A](r: Attempt[A]) extends Msg[A]
+      final case class TrySet[A](id: Long, r: Attempt[A], cb: Attempt[Boolean] => Unit) extends Msg[A]
+    }
+
+    new Concurrent[F] { self =>
+      def pure[A](a: A): F[A] = F.pure(a)
+      def handleErrorWith[A](fa: F[A])(f: Throwable => F[A]): F[A] = F.handleErrorWith(fa)(f)
+      def raiseError[A](e: Throwable): F[A] = F.raiseError(e)
+      def async[A](k: (Either[Throwable, A] => Unit) => Unit): F[A] = F.async(k)
+      def runAsync[A](fa: F[A])(cb: Either[Throwable, A] => IO[Unit]): IO[Unit] = F.runAsync(fa)(cb)
+      def flatMap[A, B](fa: F[A])(f: A => F[B]): F[B] = F.flatMap(fa)(f)
+      def tailRecM[A, B](a: A)(f: A => F[Either[A, B]]): F[B] = F.tailRecM(a)(f)
+      def liftIO[A](ioa: IO[A]): F[A] = F.liftIO(ioa)
+      def suspend[A](thunk: => F[A]): F[A] = F.suspend(thunk)
+      def unsafeRunAsync[A](fa: F[A])(f: Either[Throwable, A] => IO[Unit]): Unit =
+        F.runAsync(F.shift(fa)(ec))(f).unsafeRunSync
+
+      def race[A, B](fa: F[A], fb: F[B]): F[Either[A, B]] =
+        flatMap(ref[Either[A,B]]) { ref =>
+          flatMap(ref.race(F.map(fa)(Left(_)), F.map(fb)(Right(_)))) { _ => ref.get }
+        }
+
+      def ref[A]: F[Concurrent.Ref[F,A]] = F.delay {
+        var result: Attempt[A] = null
+        // any waiting calls to `access` before first `set`
+        var waiting: LinkedMap[MsgId, Attempt[(A, Long)] => Unit] = LinkedMap.empty
+        // id which increases with each `set` or successful `modify`
+        var nonce: Long = 0
+
+        val actor: Actor[Msg[A]] = Actor.actor[Msg[A]] {
+          case Msg.Read(cb, idf) =>
+            if (result eq null) {
+              waiting = waiting.updated(idf, cb)
+            } else {
+              val r = result
+              val id = nonce
+              ec.executeThunk { cb((r: Either[Throwable, A]).map((_,id))) }
+            }
+
+          case Msg.Set(r) =>
+            nonce += 1L
+            if (result eq null) {
+              val id = nonce
+              waiting.values.foreach { cb =>
+                ec.executeThunk { cb((r: Either[Throwable, A]).map((_,id))) }
+              }
+              waiting = LinkedMap.empty
+            }
+            result = r
+
+          case Msg.TrySet(id, r, cb) =>
+            if (id == nonce) {
+              nonce += 1L; val id2 = nonce
+              waiting.values.foreach { cb =>
+                ec.executeThunk { cb((r: Either[Throwable, A]).map((_,id2))) }
```

Comment:
Ok, I'm assuming I just wasn't following the logic correctly.  I was going off the types and some loose control flow analysis, and it seemed… odd.